docs/source/torchao_vllm_integration.md

```diff
@@ -0,0 +1,398 @@
+from torchao.quantization import Int4WeightOnlyConfig
+
+# Create quantization configuration
+quantization_config = TorchAoConfig(
+    quant_type=Int4WeightOnlyConfig(group_size=128, use_hqq=True)
+)
+
+# Load and automatically quantize the model
+model = AutoModelForCausalLM.from_pretrained(
+    "meta-llama/Llama-3.2-1B",
+    torch_dtype="auto",
+    device_map="auto",
+    quantization_config=quantization_config
+)
+
+# Save quantized model (see Serialization section below for safe_serialization details)
+model.push_to_hub("your-username/Llama-3.2-1B-int4", safe_serialization=False)
+```
+
+```{seealso}
+For more information on quantization configs, see {class}`torchao.quantization.Int4WeightOnlyConfig` and {class}`torchao.quantization.Int8WeightOnlyConfig`.
+```
+
+(serving-with-vllm)=
+### 2. Serving with VLLM
+
+```bash
+# Start VLLM server with TorchAO quantized model
+vllm serve your-username/Llama-3.2-1B-int4 \
+    --quantization torchao \
+    --dtype bfloat16 \
+```
+
+
+(adding-new-quantization-methods)=
+## Adding New Quantization Methods to VLLM
+
+(minimal-requirements-vllm)=
+### Minimal Requirements for VLLM Compatibility
+
+To make a new TorchAO quantization method work with VLLM, you need to implement minimal tensor subclass operations that support **tensor parallelism**. VLLM uses `narrow()` and `copy_()` to move data from host cpu loaded in a state dict to the device, these  require these specific aten operations:
+
+(why-these-operations)=
+### Why these ?
+
+VLLM's tensor parallelism works by:
+1. **{meth}`~torch.Tensor.narrow`** - Slicing weight tensors across different dimensions
+2. **Sharding** - Distributing tensor chunks across multiple GPUs
+3. **{meth}`~torch.Tensor.copy_`** - Moving tensor data between devices
+4. **{meth}`~torch.Tensor.detach`**
+
+
+A helpful pattern for doing this is `_apply_fn_to_data`, a method that applies a given function to all the attributes on your class w/ Tensor types. Below is a generic implementation that should work for most subclasses. We make heavy use of this, you can see an examples in {doc}`mx_ops.py <../torchao/prototype/mx_formats/mx_ops.py>`:
+
+```python
+def _apply_fn_to_data(self, fn: Callable):
+    """Applies a fn to all tensor components stored on this class"""
+    tensor_names, ctx = self.__tensor_flatten__()
+
+    # Apply the function to each tensor component
+    new_tensors = {}
+    for name in tensor_names:
+        new_tensors[name] = fn(getattr(self, name))
+
+    return self.__class__.__tensor_unflatten__(
+        new_tensors,
+        ctx,
+        None,  # outer_size parameter
+        None,  # outer_stride parameter
+    )
+```
+
+(step-by-step-guide)=
+## Step-by-Step Guide to Add a New Quantization Method
+
+(create-tensor-subclass)=
+#### 1. Create Your Tensor Subclass
+
+```python
+from torchao.core.config import AOBaseConfig
+from torchao.utils import TorchAOBaseTensor
+
+@dataclass
+class MyNewQuantConfig(AOBaseConfig):
```

Comment:
Will this config work automatically with `TorchAoConfig` or is there a hardcoded list somewhere for configs that HF knows about?

Reply:
No hard coded list